lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/session/PropertyMetadataUtil.java

```diff
@@ -68,4 +97,134 @@ public static PropertyMetadata<Duration> durationProperty(String name, String de
                 },
                 Duration::toString);
     }
+
+    public static PropertyBuilder property(String name)
+    {
+        return new PropertyBuilder(name);
+    }
+
+    public static class PropertyBuilder
+    {
+        private final String name;
+        private String description;
+        private Type sqlType;
+        private Class<?> javaType;
+        private @Nullable Object defaultValue;
+        private boolean hidden;
+
+        private PropertyBuilder(String name)
+        {
+            this.name = requireNonNull(name, "name is null");
+        }
+
+        @CanIgnoreReturnValue
+        public <C, T> PropertyBuilder fromConfig(C config, Function<C, T> getterMethodReference)
+        {
+            Class<?> configClass = config.getClass();
+            Method getter = findMethodByReference(configClass, configInstance -> {
+                //noinspection unchecked
+                T ignored = getterMethodReference.apply((C) configInstance);
+            });
+            description = findConfigDescriptionByGetter(configClass, getter)
+                    .orElseThrow(() -> new IllegalArgumentException("Config associated with %s.%s has no description".formatted(configClass.getName(), getter.getName())));
+            javaType = getter.getReturnType();
+            defaultValue = getterMethodReference.apply(config);
+
+            Class<?> boxedJavaType = Primitives.wrap(javaType);
+            if (boxedJavaType == Boolean.class) {
+                sqlType = BOOLEAN;
+            }
+            if (boxedJavaType == Integer.class) {
+                sqlType = INTEGER;
+            }
+            if (boxedJavaType == Long.class) {
+                sqlType = BIGINT;
+            }
+            if (boxedJavaType == String.class) {
+                sqlType = VARCHAR;
+            }
+
+            return this;
+        }
+
+        @CanIgnoreReturnValue
+        public PropertyBuilder hidden()
+        {
+            hidden = true;
+            return this;
+        }
+
+        public PropertyMetadata<?> build()
+        {
+            Function<Object, Object> decoder = Functions.identity();
+            Function<Object, Object> encoder = Functions.identity();
+
+            //noinspection unchecked,rawtypes
+            return new PropertyMetadata(
+                    name,
+                    description,
+                    sqlType,
+                    javaType,
+                    defaultValue,
+                    hidden,
+                    decoder,
+                    encoder);
+        }
+    }
+
+    private static <C> Method findMethodByReference(Class<C> clazz, Consumer<C> methodReference)
+    {
+        List<Method> calledMethods = new ArrayList<>(1);
+        try (DynamicType.Unloaded<C> interceptingClass = new ByteBuddy()
```

Comment:
This is smart :)

are those classes GCed now? I think yes - but please confirm.

Reply:
Classes get fried, but there is leak of locks in ClassLoader.parallelLockMap.
This is a known problem in JVM https://bugs.openjdk.org/browse/JDK-8037342 and doesn't seem to have a good solution.
In my local test i was running with just the platform class loader, but even if I create a throw-away classloader, it doesn't solve the problem. It's because the classloaders delegate to parent first, so when loading the generated class, the parent (not throw-away) gets asked whether it has it, and only then the class gets loaded in the child.
@losipiuk and I found a solution to do child-first/parent-last class loader.
This is equivalent to what we do here: https://github.com/airlift/bytecode/blob/master/src/main/java/io/airlift/bytecode/DynamicClassLoader.java#L115